protocol.go

```diff
@@ -0,0 +1,160 @@
+package exposed
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"github.com/google/flatbuffers/go"
+	"github.com/thesyncim/exposed/internal/protocol"
+	"io"
+)
+
+type request struct {
+	*protocol.Request
+	buf     *bytes.Buffer
+	builder *flatbuffers.Builder
+	sizebuf [4]byte
+}
+
+var (
+	errShortRead = errors.New("short read")
+)
+
+func newRequest(builder *flatbuffers.Builder, buf *bytes.Buffer) *request {
+	return &request{
+		builder: builder,
+		buf:     buf,
+	}
+}
+
+func (r *request) Reset() {
+	r.buf.Reset()
+	r.builder.Reset()
+}
+
+func (r *request) Set(operation uint64, payload []byte) {
+	// re-use the already-allocated Builder:
+	r.builder.Reset()
+	// create the name object and get its offset:
+	payload_position := r.builder.CreateByteString(payload)
+	// write the User object:
+	protocol.RequestStart(r.builder)
+	protocol.RequestAddOperation(r.builder, operation)
+
+	protocol.RequestAddPayload(r.builder, payload_position)
+	request_position := protocol.RequestEnd(r.builder)
+	// finish the write operations by our User the root object:
+	r.builder.Finish(request_position)
+
+	// return the byte slice containing encoded data:
+	return
+}
+
+func (r *request) WriteRequest(bw *bufio.Writer) error {
+
+	toWrite := r.builder.FinishedBytes()
+
+	binary.BigEndian.PutUint32(r.sizebuf[:], uint32(len(toWrite)))
+
+	_, err := bw.Write(r.sizebuf[:])
+	if err != nil {
+		return err
+	}
+	_, err = bw.Write(toWrite)
+
+	return err
+
+}
+
+func (r *request) ReadRequest(br *bufio.Reader) error {
+
+	_, err := br.Read(r.sizebuf[:])
+	if err != nil {
+		return err
+	}
+	size := binary.BigEndian.Uint32(r.sizebuf[:])
+	var n int64
+	n, err = r.buf.ReadFrom(io.LimitReader(br, int64(size)))
+	if err != nil {
+		return err
+	}
+	if n != int64(size) {
+		return errShortRead
+	}
+
+	r.Request = protocol.GetRootAsRequest(r.buf.Bytes(), 0)
+	return nil
+}
+
+type response struct {
+	*protocol.Response
+	builder *flatbuffers.Builder
+	buf     *bytes.Buffer
+	sizebuf [4]byte
+}
+
+func newResponse(builder *flatbuffers.Builder, buf *bytes.Buffer) *response {
+	return &response{
+		builder: builder,
+		buf:     buf,
+	}
+}
+func (r *response) Reset() {
+	r.buf.Reset()
+	r.builder.Reset()
+}
+
+func (r *response) WriteResponse(bw *bufio.Writer) error {
+
+	b := r.builder.FinishedBytes()
+	binary.BigEndian.PutUint32(r.sizebuf[:], uint32(len(b)))
+
+	_, err := bw.Write(r.sizebuf[:])
+	if err != nil {
+		return err
+	}
+	_, err = bw.Write(b)
+
+	return err
+
+}
+
+func (r *response) Set(err []byte, payload []byte) {
+	// re-use the already-allocated Builder:
+	b := r.builder
+	b.Reset()
+	// create the name object and get its offset:
+	err_position := b.CreateByteString(err)
+	payload_position := b.CreateByteString(payload)
+	// write the User object:
+	protocol.ResponseStart(b)
+	protocol.ResponseAddError(b, err_position)
+	protocol.ResponseAddPayload(b, payload_position)
+	response_position := protocol.ResponseEnd(b)
```

Comment:
don't use underscores in Go names; var response_position should be responsePosition